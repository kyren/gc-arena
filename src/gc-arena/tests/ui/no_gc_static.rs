use gc_arena::{Arena, ArenaParameters, Collect, Gc, Rootable};

fn main() {
    #[derive(Collect)]
    #[collect(no_drop)]
    struct TestRoot<'gc> {
        test: Gc<'gc, u8>,
    }

    let arena = Arena::<Rootable![TestRoot<'gc>]>::new(ArenaParameters::default(), |mc| TestRoot {
        test: Gc::new(mc, 4),
    });

    let arena = Box::leak(Box::new(arena));
    let _test: Gc<'static, _> = arena.mutate(|_, root| root.test);
}
