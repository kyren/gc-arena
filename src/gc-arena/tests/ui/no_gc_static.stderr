error: lifetime may not live long enough
  --> tests/ui/no_gc_static.rs:15:56
   |
15 |     let _test: Gc<'static, _> = arena.mutate(|_, root| root.test);
   |                                               -      - ^^^^^^^^^ returning this value requires that `'1` must outlive `'2`
   |                                               |      |
   |                                               |      return type of closure is Gc<'2, u8>
   |                                               has type `MutationContext<'1, '_>`
   |
   = note: requirement occurs because of the type `Gc<'_, u8>`, which makes the generic argument `'_` invariant
   = note: the struct `Gc<'gc, T>` is invariant over the parameter `'gc`
   = help: see <https://doc.rust-lang.org/nomicon/subtyping.html> for more information about variance
